sample-code/vc-using-framestream/vc-using-framestream/Connection.cpp

```diff
@@ -0,0 +1,148 @@
+    systemType, salinity, initialFocusRange);
+}
+
+Connection::Connection(
+  boost::asio::io_service & io,
+  boost::asio::ip::tcp::socket && commandSocket,
+  boost::asio::ip::address receiveFrom,
+  std::function<void(Aris::Network::FrameBuilder&)> onFrameCompletion,
+  SystemType systemType,
+  aris::Command::SetSalinity::Salinity salinity,
+  float initialFocusRange)
+  : commandSocket_(std::move(commandSocket))
+  , io_(io)
+  , ping_timer_(io, kPingTimerPeriod)
+  , ping_template_(std::move(CreatePingTemplate()))
+  , sendPing_([this](auto e) { this->HandlePingTimer(e); })
+  , systemType_(systemType)
+  , salinity_(salinity)
+  , onFrameCompletion_(onFrameCompletion)
+  // Note that framePartListener_ is initialized AFTER onFrameCompletion_ due to declaration
+  // order in the class declaration.
+  , frameStreamListener_(io, onFrameCompletion_, []() { return kFramePartBufferSize; }, receiveFrom)
+  , shutting_down_(false)
+{
+  assert(onFrameCompletion);
+
+  // Initialize the sonar appropriately.
+  SendCommand(CommandBuilder::SetTime());
+  SendCommand(
+    CommandBuilder::SetFrameStreamReceiver(
+      //frameStreamListener_.LocalEndpoint().address().to_string().c_str(),
+      frameStreamListener_.LocalEndpoint().port()));
+  SendCommand(
+    CommandBuilder::RequestAcousticSettings(
+      SetCookie(
+        DefaultAcousticSettingsForSystem[static_cast<unsigned>(systemType_)])));
+  SendCommand(CommandBuilder::SetSalinity(salinity));
+  SendCommand(CommandBuilder::SetFocusRange(initialFocusRange));
+
+  // Set up the connection ping timer.
+  ping_timer_.async_wait(sendPing_);
+}
+
+Connection::~Connection()
+{
+  shutting_down_ = true;
+}
+
+void Connection::HandlePingTimer(const boost::system::error_code& e) {
+  if (e || shutting_down_) {
+    return;
+  }
+
+  const void * buf = &ping_template_[0];
+  size_t size = ping_template_.capacity();
+  commandSocket_.send(boost::asio::buffer(buf, size));
+
+  ping_timer_.expires_from_now(kPingTimerPeriod);
+  ping_timer_.async_wait(sendPing_);
+}
+
+AcousticSettings Connection::SetCookie(const AcousticSettings & settings)
+{
+  auto adjustedSettings = settings;
+  adjustedSettings.cookie = cookie_.Next();
+  return adjustedSettings;
+}
+
+/* static */
+std::vector<uint8_t> Connection::CreatePingTemplate() {
+  aris::Command command = std::move(Aris::Network::CommandBuilder::Ping());
+
+  const uint32_t msgLength = command.ByteSize();
+  const auto prefixLength = sizeof msgLength;
+
+  // Make enough space to store the message length prefix with the message.
+  std::vector<uint8_t> buf(prefixLength + msgLength);
+
+  // The message is prefixed by the message length in network order.
+  *reinterpret_cast<uint32_t*>(&buf[0]) = htonl(command.ByteSize());
+  command.SerializeToArray(&buf[prefixLength], msgLength);
+
+  // Normally you would just send the command to the ARIS rather than
+  // storing it, but the ping doesn't have any content that changes.
+  return std::move(buf);
```

Comment:
Yes, best current practice seems to be no move, don't return an rvalue ref, trust the compiler.
http://stackoverflow.com/a/4986802/83202